### ISSUE TITLE
TypeError occurred in matchAll

### ISSUE THREAD

**Issue by isnot** [RomanBurunkov/tm-node-mpd#2]
_Originally posted by @RomanBurunkov in https://github.com/RomanBurunkov/tm-node-mpd/issues/1#issuecomment-775914190_

> Could you please elaborate the issue with TypeError occurred > in matchAll.
> In which conditions that happens?

[isnot] For more details, I wrote test case for this problem.
please refer my branch 'test-isnot'. https://github.com/isnot/tm-node-mpd/pull/1
In my environment, TypeError came from each time at 'mpd-on-update'.

```js
const matches = [...message.match(/(?:changed:\s*)(.*)/g)];
```

> - Match doesn't catch the groups with a /g flag so do we really need them in regex?

[isnot] If messages from MPD server constains multiple 'changed:', maybe need '/g flag'.
But I haven't actually confirmed this situation yet.

> - If there are no match it returns null so with spread operator we got an array with a one element like [null] with all consequences
>  - Since no regex groups you have to .substring(9) but in case you have null in result array you will get an error.

[isnot] I will confirm more on this point

Thank you.

**Comment by isnot** [comment]
@RomanBurunkov Notification

**Comment by RomanBurunkov** [comment]
Looks you are using an old node version, where matchAll hasn't been released.

You can use a polyfill for it or update node to 12 or higher. I believe it should be present in node 12.

See more more details about the polyfill at [MatchAll](https://www.npmjs.com/package/string.prototype.matchall)

E.g. you can use some of the match-all packages from the npm and add matchAll method to the string proto.

```js
if(!String.prototype.matchAll) {
	(function() {
		String.prototype.matchAll = function(reg) {
			return matchAll(this, reg);
		};
	})();
}
```

**Comment by isnot** [comment]
@RomanBurunkov 

I understand. I'm using Node.js v10.
Also, the polyfill you suggested will should work fine.

By the way, I will try another way.
using a simple String method instead of RegExp.

 _cf. my updated fix for [src/mpd.js](https://github.com/isnot/tm-node-mpd/commit/09a23fe303541a5f5d768619b29984f95a4c7368#diff-d09c79e9fb817bd078f20781e2f4e9328e69040b9b0f09ce50811becdebd39ffR371) and test case._

```js
  _matchAllChanged(message = '') {
    const data = [];
    if (typeof message !== 'string' || message === '') {
      return data;
    }
    for (const line of message.split('\n')) {
      if (line.indexOf('changed:') === 0) {
        data.push(line.substring(8).trim());
      }
    }
    return data;
  }
```

On the other hand, this product currently has only one "matchAll".
It should be easy to rewrite.
And it can be adapted to a wider range of use cases.

Normaly, reducing RegExp is expected to improve performance.
In some cases, it may prevent potential ReDoS vulnerabilities in the future.

I know it's an hyperbole to say about this one fix.
But it shouldn't be a bad way.


Thank you and regards.

**Comment by RomanBurunkov** [comment]
Fixed in 0.3.3
